src/Verify.MSTest/Verifier.cs

```diff
@@ -0,0 +1,215 @@
+    {
+        if (useUniqueDirectory)
+        {
+            settings.UseUniqueDirectory();
+        }
+
+        if (CurrentTestContext.Value is null)
+        {
+            throw new($"TestContext is null. {AttributeUsageHelp}");
+        }
+
+        var testName = CurrentTestContext.Value.TestName;
+        if (testName is null)
+        {
+            throw new($"TestContext.TestName is null. {AttributeUsageHelp}");
+        }
+
+        var typeName = CurrentTestContext.Value.FullyQualifiedTestClassName;
+        if (typeName is null)
+        {
+            throw new($"TestContext.FullyQualifiedTestClassName is null. {AttributeUsageHelp}");
+        }
+
+        if(!TryGetTypeFromTestContext(typeName, CurrentTestContext.Value, out var type))
+        {
+            type = FindType(typeName);
+        }
+
+        var testNameSpan = testName.AsSpan();
+        var indexOf = testNameSpan.IndexOf('(');
+        if (indexOf > 0)
+        {
+            testNameSpan = testNameSpan[..indexOf];
+        }
+
+        indexOf = testNameSpan.IndexOf('.');
+        if (indexOf > 0)
+        {
+            testNameSpan = testNameSpan[(indexOf + 1)..];
+        }
+
+        VerifierSettings.AssignTargetAssembly(type.Assembly);
+        var method = FindMethod(type, testNameSpan);
+
+        var pathInfo = GetPathInfo(sourceFile, type, method);
+        return new(
+            sourceFile,
+            settings,
+            type.NameWithParent(),
+            method.Name,
+            method.ParameterNames(),
+            pathInfo);
+    }
+
+    /// <summary>
+    /// As an optimization, try to retrieve the <see cref="ITestMethod"/> stored on the
+    /// <see cref="TestContext"/>, and use that to retrieve the <see cref="Type"/>.
+    ///
+    /// If reflection fails, return <c>false</c>.
+    /// </summary>
+    /// <param name="typeName">The fully qualified name of the class of the currently running test.</param>
+    /// <param name="testContext">The <see cref="TestContext"/> of the current test.</param>
+    /// <param name="type">The <see cref="Type"/> of the currently running test.</param>
+    /// <returns><c>true</c> if the reflection succeeded; <c>false</c> otherwise.</returns>
+    static bool TryGetTypeFromTestContext(string typeName, TestContext testContext, [NotNullWhen(true)] out Type? type)
+    {
+        try
+        {
+            // We can't use UnsafeAccessor in this context because _testMethod is of type
+            // Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.ObjectModel.TestMethod, which isn't part of the MSTest.TestFramework
+            // package. If / when an API like https://github.com/dotnet/runtime/issues/90081 lands this generic reflection can be replaced.
+            var testMethod = testContext
+                .GetType()
+                .GetField("_testMethod", BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic)
```

Comment:
can we cache this field

Reply:
The short answer is no. We can't use `typeof(TestContext)` because `TestContext` is an abstract base class and doesn't declare `_testMethod`. That field is declared on `Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.TestContextImplementation`, which isn't a public type in the MSTest.TestFramework package. Thus the need for the `.GetType()` call.

We _could_ make a cache of Type -> Field, however the runtime already has the `RuntimeTypeCache` (some info available @ https://learn.microsoft.com/en-us/archive/msdn-magazine/2005/july/using-net-avoid-common-performance-pitfalls-for-speedier-apps#S7 section "I Want My MemberInfo") which does it for us (verified by stepping into the framework code).

Let me know if I'm misunderstanding the request. Thanks!

Reply:
I moved the reflection work out of `Verifier` and into `TestContextReflector` to centralize all the reflection work in case we want to make additional changes now or in the future.